### PR TITLE
Update edition, deps, toolchain

### PR DESCRIPTION
**Pull request by crawfxrd**
Update:

- toolchain: 1.61.0 -> 1.70.0
- edition: 2018 -> 2021
- `buildchain`: 0.4.12 -> 0.5.0
- `sha2`: 0.9.9 -> 0.10.8
- `tempdir`: Replaced with `tempfile` 3.8.0
- `uuid`: 0.8.2 -> 1.4.1

Fix:

- `clippy::into_iter_on_ref`
- `clippy::needless_borrow`
- `clippy::op_ref`
- `clippy::or_fun_call`
- `clippy::redundant_clone`
- `clippy::redundant_field_names`
- `clippy::redundant_static_lifetimes`
- `clippy::search_is_some`
- `clippy::single_component_path_imports`
- `clippy::useless_format`

Ignore:

- `clippy::single_match`

### CODE REVIEW

File: rust-toolchain [hunk 1]
```diff
@@ -1 +1 @@
-1.56.1
+1.62.1
```
Comment (crawfxrd) [inline rust-toolchain:1]: Do we have some kind of minimum version? Our toolchain files seem to be all over the place.

File: Cargo.toml [hunk 2]
```diff
@@ -18,12 +18,12 @@ ecflash = { git = "https://github.com/system76/ecflash.git", branch = "stable" }
 libc = "0.2"
 plain = "0.2"
 rust-lzma = "0.5"
-serde = { version = "1.0.116", features = ["derive"] }
+serde = { version = "1.0.140", features = ["derive"] }
 serde_json = "1.0"
-sha2 = "0.9"
+sha2 = "0.10"
 tar = "=0.4.35"
```
Comment (crawfxrd) [inline Cargo.toml:24]: Need to check if tar can be updated (ref: #71).

**Review by XV-02 — DISMISSED** [review]
On both open and closed firmware systems, this successfully flashed the latest released firmware. It built correctly (though did reveal a needed tweak to the `README` instructions, namely that those looking to build should run `apt build-dep system76-firmware` first) and didn't present any regressions while testing those features. It was tested across three laptops (Pang11, Lemp9, Oryp9) and a desktop (Mira-B1)

**Review by 13r0ck — APPROVED** [review]

**Review by jackpot51 — APPROVED** [review]

**Review by jackpot51 — DISMISSED** [review]

**Review by XV-02 — DISMISSED** [review]
Everything is looking good. Tested on Galp6, Oryp10, and Galp5. Works without any issues. Approved.

**Comment by crawfxrd** [comment]
It looks  like https://github.com/alexcrichton/tar-rs/commit/d90a02f582c03dfa0fd11c78d608d0974625ae5d is what broke it. It's failing on the `res/` directory in the archive, which it's suppose to create while extracting.

File: Cargo.toml (hunk 2 above)
Comment (crawfxrd) [reply Cargo.toml]: Fixed the problem with newer versions of tar.

**Comment by crawfxrd** [comment]
- Archive timestamp issue fixed in system76/firmware/pull/476
- Issue with newer versions of tar fixed
- Replaced deprecated tempdir with tempfile

**Review by jackpot51 — DISMISSED** [review]

**Review by leviport — CHANGES_REQUESTED** [review]
This appears to break the firmware application in Ubuntu Focal. I think we probably just need to make a `master_focal` so the old version can remain on the old release.

**Comment by crawfxrd** [comment]
> This appears to break the firmware application in Ubuntu Focal.

Does it fail to start or does it not work in some way?

**Comment by leviport** [comment]
The window appears, but the window is blank and I can't interact with it, then the "force quit" prompt appears.

**Comment by crawfxrd** [comment]
I'm going to do the tar changes separately, since I think that is what is breaking it.

**Comment by leviport** [comment]
Yep, looks like that fixed it on Ubuntu Focal. I'm going to do some quick tests on Pop, but I'll probably be approving this very soon.

**Review by leviport — DISMISSED** [review]

**Review by leviport — APPROVED** [review]
Seems good on Pop and Ubuntu

**Review by jackpot51 — APPROVED** [review]

**Pull request merged**